src/server/raftkv/mod.rs

```diff
@@ -879,6 +885,37 @@ impl ReadIndexObserver for ReplicaReadLockChecker {
                         .observe(begin_instant.saturating_elapsed().as_secs_f64());
                 }
             }
+            if rctx.locked.is_none() && !start_ts.is_max() {
+                if let (Some(region_start_key), Some(region_end_key)) =
+                    (region_start_key, region_end_key)
+                {
+                    // check if there is a memory lock on a region
+                    let start_key = key_bound(region_start_key.to_vec());
+                    let end_key = key_bound(region_end_key.to_vec());
+
+                    let res = self.concurrency_manager.read_range_check(
+                        start_key.as_ref(),
+                        end_key.as_ref(),
+                        |key, lock| {
+                            // It returns immediately upon encountering a lock in a region,
+                            // regardless of the timestamp.This optimization is for the read index
+                            // cache on the follower side. Considering
+                            // timestamps might require scanning the
+                            // entire region.
+                            let raw_key = key.to_raw()?;
+                            Err(txn_types::Error::from(txn_types::ErrorInner::KeyIsLocked(
+                                lock.clone().into_lock_info(raw_key),
+                            )))
+                        },
+                    );
+                    if !matches!(
+                        res,
+                        Err(txn_types::Error(box txn_types::ErrorInner::KeyIsLocked(_)))
+                    ) {
+                        rctx.read_index_safe_ts = Some(start_ts.into_inner());
```

Comment:
It's more robust to update the `read_index_safe_ts` only when
```
if let(Ok(_)) = res {
    ...
}
```
This ensures that if the range check encounters unexpected errors other than `keyIsLocked`, the `read_index_safe_ts` won't be updated, which is safer.